src/Components/Web/src/Routing/NavLink.cs

```diff
@@ -124,9 +132,50 @@ private bool ShouldMatch(string currentUriAbsolute)
             return true;
         }
 
-        return false;
+        if (_disableMatchAllIgnoresLeftUriPart || Match != NavLinkMatch.All)
+        {
+            return false;
+        }
+
+        string uriWithoutQueryAndFragment = GetUriIgnoreQueryAndFragment(currentUriAbsolute);
+        if (EqualsHrefExactlyOrIfTrailingSlashAdded(uriWithoutQueryAndFragment))
+        {
+            return true;
+        }
+        _hrefAbsolute = GetUriIgnoreQueryAndFragment(_hrefAbsolute);
+        return EqualsHrefExactlyOrIfTrailingSlashAdded(uriWithoutQueryAndFragment);
     }
 
+    private static string GetUriIgnoreQueryAndFragment(string uri)
+    {
+        if (string.IsNullOrEmpty(uri))
+        {
+            return string.Empty;
+        }
+
+        var queryStartPos = uri.IndexOf('?');
+        var fragmentStartPos = uri.IndexOf('#');
+
+        if (queryStartPos < 0 && fragmentStartPos < 0)
+        {
+            return uri;
+        }
+
+        int minPos;
+        if (queryStartPos < 0)
+        {
+            minPos = fragmentStartPos;
+        }
+        else if (fragmentStartPos < 0)
+        {
+            minPos = queryStartPos;
+        }
+        else
+        {
+            minPos = Math.Min(queryStartPos, fragmentStartPos);
+        }
+        return uri.AsMemory(0..minPos).ToString();
```

Comment:
Can we avoid the tostring here? You can compare two spans with https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.sequenceequal?view=net-9.0#system-memoryextensions-sequenceequal-1(system-readonlyspan((-0))-system-readonlyspan((-0))-system-collections-generic-iequalitycomparer((-0)))